File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        # approach: two pointers, compute area at each iteration
        
        result = 0
        l = 0
        r = len(height) - 1
        
        while l < r:
            area = (r - l) * min(height[l], height[r])
            result = max(result, area)
            
            if height[l] > height[r]:
                r -= 1
            else:
                l += 1
        
        return result
            

File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        // approach: two pointers
        
        int result = 0;
        
        int l = 0;
        int r = height.length - 1;
        
        while (l < r) {
            int area = (r - l) * Math.min(height[l], height[r]);
            result = Math.max(result, area);
            
            if (height[r] < height[l]) {
                r--;
            } else {
                l++;
            }
        }
        
        return result;
    }
}